File: dart_sample/lib/aspect_ratio_demo_page.dart
import 'package:flutter/material.dart';

import 'counter_widgets.dart';

class AspectRatioDemoPage extends StatefulWidget {
  const AspectRatioDemoPage({super.key});

  @override
  State<AspectRatioDemoPage> createState() => _AspectRatioDemoPageState();
}

class _AspectRatioDemoPageState extends State<AspectRatioDemoPage> {
  double _aspectRatio = 16 / 9;
  int _spacerFlex = 1;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'AspectRatio + Spacer',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'AspectRatio enforces a tight ratio box; Spacer reserves remaining Row space by flex.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: '16:9',
              onTap: () => _setAspectRatio(16 / 9),
              width: 88,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: '4:3',
              onTap: () => _setAspectRatio(4 / 3),
              width: 88,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: '1:1',
              onTap: () => _setAspectRatio(1),
              width: 88,
              background: const Color(0xFFDCE3ED),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'Flex 1',
              onTap: () => _setSpacerFlex(1),
              width: 88,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'Flex 2',
              onTap: () => _setSpacerFlex(2),
              width: 88,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'Flex 3',
              onTap: () => _setSpacerFlex(3),
              width: 88,
              background: const Color(0xFFE9F5EC),
            ),
          ],
        ),
        Text(
          'ratio=${_formatRatio(_aspectRatio)}, spacerFlex=$_spacerFlex (second spacer fixed=1)',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 260,
          height: 190,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(10),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 10,
            children: <Widget>[
              AspectRatio(
                aspectRatio: _aspectRatio,
                child: Container(
                  decoration: BoxDecoration(
                    color: const Color(0xFFCCE3FF),
                    border: Border.all(color: const Color(0xFF1D3557), width: 2),
                    borderRadius: BorderRadius.circular(10),
                  ),
                  child: const Center(
                    child: Text(
                      'Aspect preview',
                      style: TextStyle(fontSize: 14, color: Colors.black),
                    ),
                  ),
                ),
              ),
              Container(
                height: 44,
                color: Colors.white,
                padding: const EdgeInsets.fromLTRB(8, 6, 8, 6),
                child: Row(
                  children: <Widget>[
                    Container(
                      width: 42,
                      height: 24,
                      color: const Color(0xFF1D3557),
                      child: const Center(
                        child: Text(
                          'L',
                          style: TextStyle(fontSize: 12, color: Colors.white),
                        ),
                      ),
                    ),
                    Spacer(flex: _spacerFlex),
                    Container(
                      width: 28,
                      height: 24,
                      color: const Color(0xFF2A9D8F),
                      child: const Center(
                        child: Text(
                          'M',
                          style: TextStyle(fontSize: 12, color: Colors.white),
                        ),
                      ),
                    ),
                    const Spacer(),
                    Container(
                      width: 54,
                      height: 24,
                      color: const Color(0xFF457B9D),
                      child: const Center(
                        child: Text(
                          'R',
                          style: TextStyle(fontSize: 12, color: Colors.white),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _setAspectRatio(double value) {
    setState(() {
      _aspectRatio = value;
    });
  }

  void _setSpacerFlex(int value) {
    setState(() {
      _spacerFlex = value;
    });
  }

  static String _formatRatio(double ratio) {
    if ((ratio - (16 / 9)).abs() < 0.001) {
      return '16:9';
    }

    if ((ratio - (4 / 3)).abs() < 0.001) {
      return '4:3';
    }

    if ((ratio - 1).abs() < 0.001) {
      return '1:1';
    }

    return ratio.toStringAsFixed(2);
  }
}

File: src/Sample/Flutter.Net/AspectRatioDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Flutter.Rendering;
using Flutter.Widgets;

// Dart parity source (reference): dart_sample/lib/aspect_ratio_demo_page.dart (exact sample parity)

namespace Flutter.Net;

public sealed class AspectRatioDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new AspectRatioDemoPageState();
    }
}

internal sealed class AspectRatioDemoPageState : State
{
    private double _aspectRatio = 16.0 / 9.0;
    private int _spacerFlex = 1;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("AspectRatio + Spacer", fontSize: 20, color: Colors.Black),
                new Text(
                    "AspectRatio enforces a tight ratio box; Spacer reserves remaining Row space by flex.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("16:9", () => SetAspectRatio(16.0 / 9.0), width: 88, colorHex: "#FFDCE3ED"),
                        BuildButton("4:3", () => SetAspectRatio(4.0 / 3.0), width: 88, colorHex: "#FFDCE3ED"),
                        BuildButton("1:1", () => SetAspectRatio(1.0), width: 88, colorHex: "#FFDCE3ED"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Flex 1", () => SetSpacerFlex(1), width: 88, colorHex: "#FFE9F5EC"),
                        BuildButton("Flex 2", () => SetSpacerFlex(2), width: 88, colorHex: "#FFE9F5EC"),
                        BuildButton("Flex 3", () => SetSpacerFlex(3), width: 88, colorHex: "#FFE9F5EC"),
                    ]),
                new Text(
                    $"ratio={FormatRatio(_aspectRatio)}, spacerFlex={_spacerFlex} (second spacer fixed=1)",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 260,
                    height: 190,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(10),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 10,
                        children:
                        [
                            new AspectRatio(
                                aspectRatio: _aspectRatio,
                                child: new Container(
                                    decoration: new BoxDecoration(
                                        Color: Color.Parse("#FFCCE3FF"),
                                        Border: new BorderSide(Color.Parse("#FF1D3557"), 2),
                                        BorderRadius: BorderRadius.Circular(10)),
                                    child: new Center(
                                        child: new Text("Aspect preview", fontSize: 14, color: Colors.Black)))),
                            new Container(
                                height: 44,
                                color: Colors.White,
                                padding: new Thickness(8, 6, 8, 6),
                                child: new Row(
                                    children:
                                    [
                                        new Container(
                                            width: 42,
                                            height: 24,
                                            color: Color.Parse("#FF1D3557"),
                                            child: new Center(
                                                child: new Text("L", fontSize: 12, color: Colors.White))),
                                        new Spacer(flex: _spacerFlex),
                                        new Container(
                                            width: 28,
                                            height: 24,
                                            color: Color.Parse("#FF2A9D8F"),
                                            child: new Center(
                                                child: new Text("M", fontSize: 12, color: Colors.White))),
                                        new Spacer(),
                                        new Container(
                                            width: 54,
                                            height: 24,
                                            color: Color.Parse("#FF457B9D"),
                                            child: new Center(
                                                child: new Text("R", fontSize: 12, color: Colors.White))),
                                    ])),
                        ])),
            ]);
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void SetAspectRatio(double value)
    {
        SetState(() => _aspectRatio = value);
    }

    private void SetSpacerFlex(int value)
    {
        SetState(() => _spacerFlex = value);
    }

    private static string FormatRatio(double ratio)
    {
        if (Math.Abs(ratio - (16.0 / 9.0)) < 0.001)
        {
            return "16:9";
        }

        if (Math.Abs(ratio - (4.0 / 3.0)) < 0.001)
        {
            return "4:3";
        }

        if (Math.Abs(ratio - 1.0) < 0.001)
        {
            return "1:1";
        }

        return ratio.ToString("0.##");
    }
}
